src/metabase/api/routes/lazy.clj

```diff
@@ -0,0 +1,75 @@
+(ns metabase.api.routes.lazy
+  "Utils for creating lazy-loading Compojure routes, for fast Metabase startup times."
+  (:require [clojure.tools.logging :as log]
+            [compojure.core :as compojure]
+            [metabase.plugins.classloader :as classloader]))
+
+(defn handler
+  "Create a lazy-loading Ring request handler. `thunk` is called on the first request, and should return a handler; this
+  value is cached and used for all subsequent requests."
+  [thunk]
+  (let [dlay (delay (thunk))]
+    (fn [request respond raise]
+      (@dlay request respond raise))))
+
+(defn ns-handler
+  "A lazy-loading Ring request handler that requires (in a thread-safe way) the namespace named by `ns-symb` and looks
+  for the var named `routes`.
+
+    (ns-handler 'metabase.api.util) -> <lazy-handler>
+
+    (<lazy-handler> request respond raise) -> (metabase.api.util/routes request respond raise)"
+  [ns-symb]
+  (handler
+   (fn []
+     (log/debugf "Lazy-loading API routes in %s" ns-symb)
+     ;; classloader/require is thread-safe
+     (classloader/require ns-symb)
+     (ns-resolve ns-symb 'routes))))
```

Comment:
In the extreme performance edition I did a `var-get` here so we don't have to do var lookup for each request. Profiling indicates that the performance difference is minimal. By returning the var here, the code is REPL friendly in the sense that reloading an API namespace won't require you to reload `metabase.api.routes` as well.